homeassistant/components/mqtt/__init__.py

```diff
@@ -526,36 +520,54 @@ def stop():
         return self.hass.async_add_job(stop)
 
     @asyncio.coroutine
-    def async_subscribe(self, topic, qos):
-        """Subscribe to a topic.
+    def async_subscribe(self, topic, msg_callback, qos, encoding):
+        """Set up a subscription to a topic with the provided qos.
 
         This method is a coroutine.
         """
         if not isinstance(topic, str):
-            raise HomeAssistantError("topic need to be a string!")
+            raise HomeAssistantError("topic needs to be a string!")
 
-        with (yield from self._paho_lock):
-            if topic in self.subscribed_topics:
+        subscription = Subscription(topic, msg_callback, qos, encoding)
+        self.subscriptions.append(subscription)
+
+        yield from self._async_perform_subscription(topic, qos)
+
+        @callback
+        def async_remove():
+            """Remove subscription."""
+            if subscription not in self.subscriptions:
+                raise HomeAssistantError("Can't remove subscription twice")
+            self.subscriptions.remove(subscription)
+
+            grouped = self._grouped_subscriptions()
+            if topic in grouped:
+                # Other subscriptions on topic remaining - don't unsubscribe.
                 return
-            self.wanted_topics[topic] = qos
-            result, mid = yield from self.hass.async_add_job(
-                self._mqttc.subscribe, topic, qos)
+            self.hass.async_add_job(self._async_unsubscribe(topic))
 
-            _raise_on_error(result)
-            self.progress[mid] = topic
+        return async_remove
 
     @asyncio.coroutine
-    def async_unsubscribe(self, topic):
-        """Unsubscribe from topic.
+    def _async_unsubscribe(self, topic):
+        """Unsubscribe from a topic.
 
         This method is a coroutine.
         """
-        self.wanted_topics.pop(topic, None)
-        result, mid = yield from self.hass.async_add_job(
-            self._mqttc.unsubscribe, topic)
+        with (yield from self._paho_lock):
```

Comment:
This could potentially result in a race condition - so better safe than sorry and acquire the lock.